### PR TITLE
fix: fix a bug in c apis and add more tests 

### PR DESCRIPTION
**Pull request by yangby-cryptape**
- [x] refactor(tests): split tests generators
- [x] fix(bindings/c): offsets are incorrect when all items in a Vector are None
- [x] test: test all test vectors in C apis
- [x] test: add proptest for both Rust and C

### CODE REVIEW

**Comment by yangby-cryptape** [comment]
Fix the follow test cases in C:
https://github.com/nervosnetwork/molecule/blob/a1eaca39ffbb210a118c34cdfba7c41832b108a6/test/vectors/simple.yaml#L263-L280

File: bindings/c/include/molecule_builder.h [hunk 1]
```diff
@@ -262,15 +262,13 @@ MOLECULE_API_DECORATOR mol_seg_res_t mol_dynvec_builder_finalize(mol_builder_t b
     res.seg.size = MOL_NUM_T_SIZE + builder.number_used + builder.data_used;
     res.seg.ptr = (uint8_t*)malloc(res.seg.size);
     mol_pack_number(res.seg.ptr, &res.seg.size);
-    if (builder.data_used > 0) {
```
Comment (yangby-cryptape) [inline bindings/c/include/molecule_builder.h]: When N `None` were pushed into a Vector (`N >= 1`), the data_used is still `0`, but N offsets have to be saved.

**Comment by xxuejie** [comment]
This is not a hard rule but next time I would suggest we split the test part and the bug fix into 2 PRs. So when I'm looking at a bug fix, I know exactly what the bug covers, while mixing 2 tasks in one PR like this makes it harder to read.

**Comment by yangby-cryptape** [comment]
> This is not a hard rule but next time I would suggest we split the test part and the bug fix into 2 PRs. So when I'm looking at a bug fix, I know exactly what the bug covers, while mixing 2 tasks in one PR like this makes it harder to read.

OK, I'll pay attention next time.

I found the bug when I added tests. So I put them into one PR. The tests can approve the fix works.

For current PR, I split the commits for different sub-tasks, combine all sub-tasks is the main target: add more tests for C.
You can only review [the commit for the fix](https://github.com/nervosnetwork/molecule/pull/24/commits/f714c99905ff148f6088f905c6579e78aa7d0995).

**Comment by doitian** [comment]
🚢

**Pull request merged**